### DISCUSSION TITLE
Why is there only a deployer.phar?

### DISCUSSION THREAD

**Discussion by flolanger** [deployphp/deployer#3087] [General]
In RC4, the whole project including `bin/dep` existed. Now since RC5 only the `deployer.phar` is available.

CI Pipelines using `dep` don't work any more because of that. Will this be available again in the future, or will this be the new standard?

Thanks in advance!

**Comment by antonmedv** [comment]
This is a way I decided to try. It was previously in deployer/dist package for a couple of years. Now I decided to merge it. 

What kind of problem are you facing? Other classes not loading? Autolaod issues? This is work in process bug.

**Reply by flolanger** [reply]
The problem we're facing is that we can't run our CI Pipelines anymore, because `dep` is not available anymore globally. We have a docker based gitlab runner which has an image that executes `composer global require deployer/deployer:@beta`.

If the script then wants to execute the deployment, the following error appears:
```
$ dep -f./deployment/deploy.php deploy develop -vvv
/bin/bash: line 132: dep: command not found
```

So it would be nice to also have the `dep` command globally again and not just a deployer.phar. :)

Thanks in advance!

**Reply by antonmedv** [reply]
Reason I decided keep only deployer.phar as its allows autocomplete in phpstorm. 

Also I’m planning to create my own docker images for Deployer.

**Reply by flolanger** [reply]
I see, but can't you provide both? I think we are not the only one with this problem.

Would be great. Also, I think if dep isn't working anymore the whole docs need to be updated as well. Don't think that's worth it tbh.

**Reply by antonmedv** [reply]
I updated docs al.

**Reply by flolanger** [reply]
Can't you provide both? phar and source? So we can continue using deployer :)

**Reply by antonmedv** [reply]
You still can. Just create and alias. As in docs: 
```bash
alias dep='vendor/bin/deployer.phar'
```
Replacing source with phar allows to drop composer requirements - no more conflicting packages.

**Reply by lsv** [reply]
> Replacing source with phar allows to drop composer requirements - no more conflicting packages.

But isn't that just a bad excuse to not keep the requirements up to date?

Also its totally possible to use `composer -d anotherdir require deployer/deployer`

When you are working with phpstorm on windows, but in a WSL envorienment, deployer.phar, becomes deployer.phar.bat and then phpstorm can not auto complete anymore.

**Comment by antonmedv** [comment]
> But isn't that just a bad excuse to not keep the requirements up to date?

Nope. There is dozens of project what depend on symfony 5 components. 

> becomes deployer.phar.bat and then phpstorm can not auto complete anymore.

Original phar still there In vendors dir. only bin dir has bat file.

**Comment by havvg** [comment]
I like this very much. I consider Deployer a tool, but a like it to be installed with the project rather than globally. The change regarding the `.phar` is nice, imho. In one of our smaller projects it reduces dependencies of the project massively.

```
$ composer require -w --dev deployer/deployer:^7.0@dev

Changelogs summary:

 - symfony/yaml removed (installed version was v5.2.5)

 - symfony/string removed (installed version was v5.2.6)

 - symfony/service-contracts removed (installed version was v2.4.0)

 - symfony/process removed (installed version was v5.2.4)

 - symfony/polyfill-php80 removed (installed version was v1.22.1)

 - symfony/polyfill-php73 removed (installed version was v1.22.1)

 - symfony/polyfill-intl-normalizer removed (installed version was v1.22.1)

 - symfony/polyfill-intl-grapheme removed (installed version was v1.22.1)

 - symfony/deprecation-contracts removed (installed version was v2.4.0)

 - symfony/console removed (installed version was v5.2.6)

 - psr/container removed (installed version was 1.1.1)

 - pimple/pimple removed (installed version was v3.4.0)

 - deployer/phar-update removed (installed version was v2.2.0)

 - deployer/deployer updated from v6.8.0 to v7.0.0-rc.8 major
   See changes: https://github.com/deployphp/deployer/compare/v6.8.0...v7.0.0-rc.8
   Release notes: https://github.com/deployphp/deployer/releases/tag/v7.0.0-rc.8
```

**Comment by dustingraham** [comment]
I'm still reading through issues and discussions but the problem I have is that phpStorm does not understand deployer.

![image](https://user-images.githubusercontent.com/1414400/166327116-4d9a249f-387a-4663-af9a-c63dd100fb00.png)

I couldn't find anything in the documentation yet, but it would be nice if the files were available somewhere that phpStorm could index so that it could highlight and hint correctly.

**Reply by antonmedv** [reply]
Make sure what your PhpStorm setting support files from phar.

**Reply by dustingraham** [reply]
Thank you. Yes, phpStorm is set up to support files from phar. I see now that it is doing some inspection. For example, with the `namespace Deployer;` it understands the `set('...', '...');` function call. But, then it is confused thinking that it should be Deployer\string. In any case, I don't want to waste your time. I am disabling all inspections and just going to live without some of the completion and inspection capabilities.

![image](https://user-images.githubusercontent.com/1414400/166337667-c2ca55ed-ea6c-4291-b153-e558bff44de3.png)

**Reply by ScvWebFire** [reply]
> Thank you. Yes, phpStorm is set up to support files from phar. I see now that it is doing some inspection. For example, with the `namespace Deployer;` it understands the `set('...', '...');` function call. But, then it is confused thinking that it should be Deployer\string. In any case, I don't want to waste your time. I am disabling all inspections and just going to live without some of the completion and inspection capabilities.
> 
> ![image](https://user-images.githubusercontent.com/1414400/166337667-c2ca55ed-ea6c-4291-b153-e558bff44de3.png)

You need to set your PHP Languange level to 7.x at least. File > Settings > PHP > PHP language level

**Reply by dustingraham** [reply]
Thanks @ScvWebFire changing the language level to php 7.x+ **fixed** the parameter type!

Last thing is the require can't find the recipe path.

![image](https://user-images.githubusercontent.com/1414400/167672162-ad1da4b1-4baa-43a1-be81-c6e8356eb0ba.png)

(The yellow line under the set is a good warning because I have not wrapped the exception in try/catch.)

**Reply by leevigraham** [reply]
@dustingraham Did you find a working solution for recipes?

**Reply by dustingraham** [reply]
@leevigraham No. I contacted phpstorm and they said it is not supported.

They only suggested suppressing it by adding to the previous line: ` /** @noinspection PhpIncludeInspection */`

**Reply by antonmedv** [reply]
Other possibility:

    import(recipes/laravel.php);

**Reply by leevigraham** [reply]
@dustingraham bummer… The click to go to file is the bit I really wanted.

**Reply by dustingraham** [reply]
@leevigraham Absolutely, me too.
@antonmedv That does get around the phpstorm error message, but doesn't solve the phpstorm IDE support. Thank you for the idea.
